### PR TITLE
Refactor Mock newResource and call to accept property bag rather than individual args

### PR DESCRIPTION
**Pull request by stack72**
Fixes: #6617

@justinvp / @komalali / @lukehoban tagging you fine folks on this PR - this is the follow up to the work of https://github.com/pulumi/pulumi/pull/6551 where we want to allow extensibility in the future without any need to break things

### CODE REVIEW

**Comment by stack72** [comment]
Need to do some more work for this to be ready for merging!

File: sdk/nodejs/runtime/mocks.ts [hunk 1]
```diff
@@ -114,13 +167,14 @@ export class MockMonitor {
...
-                req.getType(),
-                req.getName(),
-                deserializeProperties(req.getObject()),
-                req.getProvider(),
-                req.getImportid(),
-                req.getCustom());
+            const result = this.mocks.newResource({
+                type: req.getType(),
+                name: req.getName(),
+                inputs: deserializeProperties(req.getProperties()),
```
Comment (EvanBoyle) [inline sdk/nodejs/runtime/mocks.ts:173]: Shoudl this be `deserializeProperties(req.getObject()),`? Does `req.getProperties` exist for register resource?

Comment (EvanBoyle) [reply sdk/nodejs/runtime/mocks.ts:173]: There might be a similar issue with `getId` vs `getImportid`, but not sure.

**Review by justinvp — COMMENTED** [review]
Have we considered doing this in a non-breaking way? Keep the existing interface as-is, introduce a new interface, and support passing both the old and new interfaces when setting mocks? Existing user code then wouldn't need to change, and we wouldn't have to get this in for 3.0.

The existing interface name is `Mocks`/`IMocks`, I could imagine coming up with a similarly decent enough name for the new interface (e.g. `TestMocks`/`ITestMocks`).

For TS/JS, we could make no change and continue to add additional parameters to the function (like we did for `custom`). Or introduce the new interface with differing function names. Or introduce an [abstract class](https://www.typescriptlang.org/docs/handbook/2/classes.html#abstract-classes-and-members) (instead of an interface) that we could runtime type-check for when the mock is set.

**Comment by stack72** [comment]
> Have we considered doing this in a non-breaking way? Keep the existing interface as-is, introduce a new interface, and support passing both the old and new interfaces when setting mocks? Existing user code then wouldn't need to change, and we wouldn't have to get this in for 3.0.
> 
> The existing interface name is `Mocks`/`IMocks`, I could imagine coming up with a similarly decent enough name for the new interface (e.g. `TestMocks`/`ITestMocks`).
> 
> For TS/JS, we could make no change and continue to add additional parameters to the function (like we did for `custom`). Or introduce the new interface with differing function names. Or introduce an [abstract class](https://www.typescriptlang.org/docs/handbook/2/classes.html#abstract-classes-and-members) (instead of an interface) that we could runtime type-check for when the mock is set.

That approach would work fine for TS but it would be much more messy for Go - this new way of thinking uses the 3.0 approach to make the break so that we have a consistent feel to the API here across all of the language SDKS

**Comment by justinvp** [comment]
> it would be much more messy for Go

Is it much more messy? This way, users don't have to modify their existing code, and we could decide to mark the existing interface and function as deprecated (and potentially remove it in a future release after giving users fair warning).

```go
// Existing (keep)...

type MockResourceMonitor interface {
	Call(token string, args resource.PropertyMap, provider string) (resource.PropertyMap, error)
	NewResource(typeToken, name string, inputs resource.PropertyMap,
		provider, id string) (string, resource.PropertyMap, error)
}

func WithMocks(project, stack string, mocks MockResourceMonitor) RunOption {
	return func(info *RunInfo) {
		info.Project, info.Stack, info.Mocks = project, stack, mocks
	}
}

// New...

type TestMockResourceMonitor interface {
	Call(args MockCallArgs) (resource.PropertyMap, error)
	NewResource(args MockResourceArgs) (string, resource.PropertyMap, error)
}

func WithTestMocks(project, stack string, mocks TestMockResourceMonitor) RunOption {
	return func(info *RunInfo) {
		info.Project, info.Stack, info.TestMocks = project, stack, mocks
	}
}
```

File: sdk/nodejs/tests_with_mocks/mocks.spec.ts [hunk 2]
```diff
@@ -37,12 +38,12 @@ pulumi.runtime.setMocks({
                     publicDns: "ec2-203-0-113-12.compute-1.amazonaws.com",
                     publicIP: "203.0.113.12",
                 };
-                return { id: "i-1234567890abcdef0", state: { ...inputs, ...state } };
+                return { id: "i-1234567890abcdef0", state: { ...args.inputs, ...state } };
             case "pkg:index:MyCustom":
-                assert.strictEqual(custom, true);
-                return { id: name + "_id", state: inputs };
+                assert.strictEqual(args.custom, true);
+                return { id: name + "_id", state: args.inputs };
```
Comment (komalali) [inline sdk/nodejs/tests_with_mocks/mocks.spec.ts:44]: This should fix the promise leak
```suggestion
                return { id: args.name + "_id", state: args.inputs };
```

Comment (stack72) [reply sdk/nodejs/tests_with_mocks/mocks.spec.ts]: 🥇

**Review by komalali — APPROVED** [review]

**Review by komalali — COMMENTED** [review]
Hmm looks like the python tests are failing from changes here.

**Comment by stack72** [comment]
> Hmm looks like the python tests are failing from changes here.

Yeah I will fix these up - this is an issue where Pulumi.runtime.Mocks is being used in another place but I didn’t update that in the tests

will get it fixed!

**Comment by stack72** [comment]
Ok, I spoke to @leezen about this - we are going to proceed with the planned approach in this PR - we have the opportunity to make a breaking change - we are going to take this opportunity here. We could indeed have gone with the suggestion above regarding keeping it non-breaking and introduce a new interface and deprecate the old one but we decided lets make the clean break here and not manage 2 code paths in the future

**Comment by adriantorrie** [comment]
Just leaving this here incase anyone else needs pointing to the new mocks for python tests if your CI/CD unit testing has failed.

https://www.pulumi.com/docs/guides/testing/unit/#add-mocks

**Old** way:

```python
class Mock(pulumi.runtime.Mocks):

    def call(self, token, args, provider):
        {}

    def new_resource(self, type_, name, inputs, provider, id_):
        # Generic return if no mocks-by-type detected
        return (name, inputs)
```

**New** way:

```python
class Mock(pulumi.runtime.Mocks):

    def call(self, pulumi.runtime.MockCallArgs):
        {}

    def new_resource(self, pulumi.runtime.MockCallArgs):
        # Then use `args.<arg_name>` when referencing, e,g.
        return [args.name + '_id', args.inputs]
```

**Pull request merged**